lib/routes/oauth.js

```diff
@@ -55,6 +58,82 @@ module.exports = (log, config, oauthdb) => {
         return oauthdb.getScopedKeyData(sessionToken, request.payload)
       }
     },
+    {
+      method: 'POST',
+      path: '/oauth/authorization',
+      options: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: oauthdb.api.createAuthorizationCode.opts.validate.payload.keys({
+            assertion: Joi.forbidden()
+          })
+        },
+        response: {
+          schema: oauthdb.api.createAuthorizationCode.opts.validate.response
+        }
+      },
+      handler: async function (request) {
+        const sessionToken = request.auth.credentials
+        return oauthdb.createAuthorizationCode(sessionToken, request.payload)
+      }
+    },
+    {
+      method: 'POST',
+      path: '/oauth/token',
+      options: {
+        auth: {
+          // XXX TODO: This route must also be able to accept 'client_secret' as Basic Auth in header.
+          mode: 'optional',
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: Joi.alternatives().try(
+            oauthdb.api.grantTokensFromAuthorizationCode.opts.validate.payload,
+            oauthdb.api.grantTokensFromRefreshToken.opts.validate.payload,
+            oauthdb.api.grantTokensFromAssertion.opts.validate.payload.keys({
+              assertion: Joi.forbidden()
+            })
+          )
+        },
+        response: {
+          schema: Joi.alternatives().try(
+            oauthdb.api.grantTokensFromAuthorizationCode.opts.validate.response,
+            oauthdb.api.grantTokensFromRefreshToken.opts.validate.response,
+            oauthdb.api.grantTokensFromAssertion.opts.validate.response
+          )
+        }
+      },
+      handler: async function (request) {
+        const sessionToken = request.auth.credentials
+        let grant
+        switch (request.payload.grant_type) {
+          case 'authorization_code':
+            grant = await oauthdb.grantTokensFromAuthorizationCode(request.payload)
+            break
+          case 'refresh_token':
+            grant = await oauthdb.grantTokensFromRefreshToken(request.payload)
+            break
+          case 'fxa-assertion':
+            if (! sessionToken) {
+              throw error.invalidToken()
+            }
+            grant = await oauthdb.grantTokensFromSessionToken(sessionToken, request.payload)
+            break
+          default:
+            throw error.internalValidationError()
+        }
+
+        switch (request.payload.grant_type) {
+          case 'authorization_code':
+          case 'fxa-assertion':
+            await oauthRouteUtils.newTokenNotification(db, oauthdb, mailer, devices, request, sessionToken, grant)
```

Comment:
Rather than switching on `grant_type`, I wonder if what you want here is to switch on whether a `refresh_token` was generated.

Reply:
`grant.refreshToken`